File: src/main/java/bai/deirn/fsn/registry/BlockEntityTypeRegistry.java
package bai.deirn.fsn.registry;

import bai.deirn.fsn.Utils;
import bai.deirn.fsn.block.entity.*;
import net.minecraft.block.entity.BlockEntityType;

public abstract class BlockEntityTypeRegistry {

    public static final BlockEntityType<?> MASTER = Utils.createBlockEntity(MasterBlockEntity::new, BlockRegistry.MASTER);
    public static final BlockEntityType<?> CRAFTING_TERMINAL = Utils.createBlockEntity(CraftingTerminalBlockEntity::new, BlockRegistry.CRAFTING_TERMINAL);
    public static final BlockEntityType<?> PROCESSING_TERMINAL = Utils.createBlockEntity(ProcessingTerminalBlockEntity::new, BlockRegistry.PROCESSING_TERMINAL);

    public static final BlockEntityType<?> CABLE = Utils.createBlockEntity(CableBlockEntity::new, BlockRegistry.CABLE);
    public static final BlockEntityType<?> STORAGE_CABLE = Utils.createBlockEntity(StorageCableBlockEntity::new, BlockRegistry.STORAGE_CABLE);

    public static void init(){
        Utils.register("master", MASTER);
        Utils.register("crafting_terminal", CRAFTING_TERMINAL);
        Utils.register("processing_terminal", PROCESSING_TERMINAL);
        Utils.register("cable", CABLE);
        Utils.register("storage_cable", STORAGE_CABLE);
    }

}

File: src/main/kotlin/badasintended/slotlink/registry/BlockEntityTypeRegistry.kt
package badasintended.slotlink.registry

import badasintended.slotlink.block.ModBlock
import badasintended.slotlink.block.entity.*
import net.minecraft.block.entity.BlockEntity
import net.minecraft.block.entity.BlockEntityType
import net.minecraft.util.registry.Registry
import net.minecraft.util.registry.Registry.BLOCK_ENTITY_TYPE
import java.util.function.Supplier

typealias T = BlockEntityType<*>
typealias B = BlockRegistry

object BlockEntityTypeRegistry {

    lateinit var MASTER: T
    lateinit var REQUEST: T

    lateinit var CABLE: T
    lateinit var LINK_CABLE: T
    lateinit var IMPORT_CABLE: T
    lateinit var EXPORT_CABLE: T

    fun init() {
        MASTER = r(B.MASTER, ::MasterBlockEntity)
        REQUEST = r(B.REQUEST, ::RequestBlockEntity)

        CABLE = r(B.CABLE, ::CableBlockEntity)
        LINK_CABLE = r(B.LINK_CABLE, ::LinkCableBlockEntity)
        IMPORT_CABLE = r(B.IMPORT_CABLE, ::ImportCableBlockEntity)
        EXPORT_CABLE = r(B.EXPORT_CABLE, ::ExportCableBlockEntity)
    }

    private fun <E : BlockEntity> r(block: ModBlock, function: () -> E): BlockEntityType<E> {
        return Registry.register(
            BLOCK_ENTITY_TYPE, block.id, BlockEntityType.Builder.create(Supplier(function), block).build(null)
        )
    }

}
